design/external-authorization-http-support-design.md

```diff
@@ -0,0 +1,165 @@
+# Global HTTP External Authorization
+
+Status: Draft
+
+## Abstract
+
+HTTP Support for External Authorization.
+
+## Background
+See [#4954](https://github.com/projectcontour/contour/issues/4954) for context.
+
+## Goals
+- Support Envoy's L7 External Authorization filter for HTTP Virtual Hosts.
+- Extend the functionality of existing External Authorization.
+
+## Non Goals
+- Supporting more than one External Authorization Configuration for HTTP endpoints.
+
+## High-Level Design
+
+Contour will add HTTP support for Envoy's External Authorozation.
+
+Currently, Contour supports External Authorization only over HTTPS. External Authorization allows for requests to be authenticated before they are forwarded upstream. However, External authorization over HTTPS is incompatible with setups that terminate TLS at the cloud provider's load balancer layer because the Envoy HCM relies on the SNI to choose the correct filter chain. 
+
+A new type, `GlobalHTTPExtAuth`, will be defined as part of the `ContourConfiguration`.
+A `GlobalHTTPExtAuth` config would define a global external authorization configuration for all the HTTP routes. 
+
+## Detailed Design
+
+### Contour Configuration changes
+By using global external authorization, an external authorization service can be configured in the Contour config file.
+This External authorization configuration will be used for all HTTP routes. 
```

Comment:
One use case we should consider is whether individual HTTP vhosts should be able to opt out of ExtAuth, if it's enabled globally. This is a little tricky since different orgs may view this functionality differently -- is it defining a global **requirement** that all HTTP vhosts must use a given ExtAuth server, or a global **default** ExtAuth config that can be opted out of by individual sites?

Personally I think I lean towards the latter, particularly since with TLS vhosts, it's up to the individual root HTTPProxy owner to decide whether to enable ExtAuth or not. And, if we move to using the matching API to enable per-HTTP vhost ExtAuth config, then we'd again support opting into ExtAuth on a per-vhost basis.

If we _do_ want to support the use case of opting out of the global config, perhaps we could do something like:

```yaml
apiVersion: projectcontour.io/v1
kind: HTTPProxy
metadata:
  name: opt-out-of-ext-auth
spec:
  virtualhost:
    fqdn: local.projectcontour.io
    authorization:
      disabled: true
```

Reply:
I see your point; it would be nice to have the flexibility to opt-out if a service owner doesn't want auth. 
This should be achievable by doing it on a per-route basis, but that could get tedious if there are a bunch of them.

Reply:
I have updated the design and added two toggles to disable auth; per route and per virtual host